### DISCUSSION TITLE
Error parsing calendar - Unexpected end of file

### DISCUSSION THREAD

**Discussion by comfy332** [bitfireAT/icsx5#67] [Q&A]
Good morning,

Trying to import a webcal, the app fails to parse the calendar:

> Couldn't parse iCalendar
> 
> at.bitfire.ical4android.InvalidCalendarException: Couldn't parse iCalendar
> 	at at.bitfire.ical4android.ICalendar$Companion.fromReader(ICalendar.kt:81)
> 	at at.bitfire.ical4android.Event$Companion.eventsFromReader(Event.kt:81)
> 	at at.bitfire.icsdroid.ui.AddCalendarValidationFragment$ValidationModel$initialize$downloader$1.onSuccess(AddCalendarValidationFragment.kt:115)
> 	at at.bitfire.icsdroid.CalendarFetcher.fetchNetwork$icsx5_62_2_0_2_standardRelease(CalendarFetcher.kt:155)
> 	at at.bitfire.icsdroid.CalendarFetcher.onRedirect(CalendarFetcher.kt:83)
> 	at at.bitfire.icsdroid.CalendarFetcher.fetchNetwork$icsx5_62_2_0_2_standardRelease(CalendarFetcher.kt:171)
> 	at at.bitfire.icsdroid.CalendarFetcher.run(CalendarFetcher.kt:46)
> 	at java.lang.Thread.run(Thread.java:920)
> Caused by: net.fortuna.ical4j.data.ParserException: Error at line 3,082:Unexpected end of file
> 	at net.fortuna.ical4j.data.CalendarParserImpl.nextToken(CalendarParserImpl.java:657)
> 	at net.fortuna.ical4j.data.CalendarParserImpl.access$1500(CalendarParserImpl.java:56)
> 	at net.fortuna.ical4j.data.CalendarParserImpl$PropertyListParser.parse(CalendarParserImpl.java:232)
> 	at net.fortuna.ical4j.data.CalendarParserImpl.parseCalendar(CalendarParserImpl.java:128)
> 	at net.fortuna.ical4j.data.CalendarParserImpl.parseCalendarList(CalendarParserImpl.java:194)
> 	at net.fortuna.ical4j.data.CalendarParserImpl.parse(CalendarParserImpl.java:163)
> 	at net.fortuna.ical4j.data.CalendarBuilder.build(CalendarBuilder.java:197)
> 	at net.fortuna.ical4j.data.CalendarBuilder.build(CalendarBuilder.java:185)
> 	at at.bitfire.ical4android.ICalendar$Companion.fromReader(ICalendar.kt:79)
> 	... 7 more

The calendar itself is a webcal, it ends with:

> TRANSP:OPAQUE END:VEVENT END:VCALENDAR

However it does not actually download an .ics file, but both Thunderbird and Google Calendar have no problem with importing it.

**Comment by rfc2822** [comment]
Hi,

Can you provide an URL to the Webcal?

**Reply by comfy332** [reply]
Not publicly, I will email it to the email in the bitfireAT account header.

**Reply by rfc2822** [reply]
(Updated)

Thanks for the file. At the end, there is:

```
TRANSP:OPAQUE
END:VEVENT

<space><space>
END:VCALENDAR
```

The space sign breaks the calendar. I think the [folding](https://datatracker.ietf.org/doc/html/rfc5545#section-3.1) is broken. Unfolded it's

```
TRANSP:OPAQUE
END:VEVENT
<space>
END:VCALENDAR
```

And then the line with space is invalid. To be honest, I haven't seen files with empty content lines at all, and I don't think they're even valid, which could be the main reason of the problem. The file seems to be generated by `PRODID:-//Scientia Ltd//iCalendar Server v 1.0//EN`… we could ask there.

(BTW, I still remember the Algorithms and Data Structures lecture 😀)

**Reply by comfy332** [reply]
Thank you so much, I can try and speak to the IT support and see what software they are using.
What makes ICSx5 fail but Thunderbird and Google Calendar have no problem?
(Edit: spelling)

**Reply by rfc2822** [reply]
ICSx⁵ uses the [ical4j](https://github.com/ical4j/ical4j/) library, which is commonly used in the Java environment, but Thunderbird and Google Calendar are not Java-based.

If you could speak to the IT support it would be nice. I'll report it to ical4j, too.

**Reply by rfc2822** [reply]
I have seen there are two spaces and this breaks the unfolding process.

I'd ask them to remove all the white space (empty lines, spaces at the end of lines) etc. to produce compliant iCalendar files.

Unfortunately we can't improve processing of such invalid files easily.
